### PR TITLE
Cover more cases in SubstrDynamicReturnTypeExtension

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/5441

see https://3v4l.org/diXok

### CODE REVIEW

File: tests/PHPStan/Analyser/data/substr-php8.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+namespace Substr;
+
+use function PHPStan\Testing\assertType;
+
+class Substr
+{
+	public function doFoo(string $s, int $i)
+	{
+		assertType('string', substr(null, true));
+		assertType('string', substr('a', 2));
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/substr-php8.php:12]: this cases always return an empty-string.. AFAIR there is no `empty-string` type, right?

**Comment by staabm** [comment]
oh my, this seem to trigger a lot of errors across the phpstan-src codebase.

I will leave it like it is for now, and wait for a initial review - so it won't get even bigger and even harder to review.

File: src/Type/Php/SubstrDynamicReturnTypeExtension.php [hunk 2]
```diff
@@ -53,7 +108,28 @@ public function getTypeFromFunctionCall(
...
+		}
+
+		// happy path
+		if ($this->phpVersion->getVersionId() >= 80000) {
+		    return new StringType();
+	    }
+
+		return TypeCombinator::union(
+			new StringType(),
+			new ConstantBooleanType(false)
```
Comment (staabm) [inline src/Type/Php/SubstrDynamicReturnTypeExtension.php:131]: IMO this is the only controversial case: returning `string|false` leads to ergonimc issues in substr() calling code.
Most codebases I know handle substr() like weakly typed - they don‘t handle a possible false-return.

we also see in the phpstan-src codebase its handled like that

**Comment by staabm** [comment]
Do you think its worse pursuing all other cases here besides the return-false case?
(E.g. constant string resolving etc?)

**Comment by ondrejmirtes** [comment]
We could cover:

1) Totally constant uses like `substr('foo', 0, 1)` where all the args are constant (in this case leading to `'f'`)
2) Non-empty-string uses like `substr($nonEmpty, 0, $x)` (leading to `non-empty-string`)

I can't think of any other.

**Pull request closed without merging**